flatdata-rs/lib/src/rawdata.rs

```diff
@@ -0,0 +1,35 @@
+/// Exposes blocks of raw data, providing auxiliary functionality like
+/// extracting substrings
+#[derive(Debug, Clone, Copy)]
+pub struct RawData<'a> {
+    data: &'a [u8],
+}
+
+impl<'a> std::ops::Deref for RawData<'a> {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        self.data
+    }
+}
+
+impl<'a> RawData<'a> {
+    /// Create a new object from raw memory reference
+    pub fn new(data: &'a [u8]) -> Self {
+        Self { data }
+    }
+
+    /// Read a \0 terminated substring starting at specified offset
+    pub fn substring(&self, start: usize) -> Result<&str, std::str::Utf8Error> {
+        let suffix = &self.data[start..];
+        match suffix.iter().position(|&c| c == 0) {
```

Comment:
This is doing 2 scans. First for searching the end and second for checking if it is a well formed utf8 string. Right now (unfortunately) `CStr` is doing the same, however they have a comment that they want to improve this. Probably, we should just use `CStr::from_ptr(..).to_str()` here to benefit from planned changes in the future.

Cf. https://doc.rust-lang.org/src/std/ffi/c_str.rs.html#1159-1165

Reply:
This would fail horribly if the string is not 0-terminated, wouldn't it?

Reply:
CStr::from_ptr checks for 0-termination at least now.

Reply:
And our buffer might not contain one. The current version handles that gracefully.